docs/common-github-actions-practices.md

```diff
@@ -0,0 +1,387 @@
+# Common GitHub Actions practices
+
+For now this is documenting desired state across the three Java repositories.
+
+Once we agree and implement, will share more broadly across OpenTelemetry
+(and surely learn and make more changes).
+
+<details>
+<summary>Table of Contents</summary>
+
+<!-- toc -->
+
+- [Have a single required status check for PRs](#have-a-single-required-status-check-for-prs)
+- [Configure "cancel-in-progress" on PR workflows](#configure-cancel-in-progress-on-pr-workflows)
+- [Prefer `gh` cli over third-party github actions for simple tasks](#prefer-gh-cli-over-third-party-github-actions-for-simple-tasks)
+- [Use github action cache to make builds faster and less flaky](#use-github-action-cache-to-make-builds-faster-and-less-flaky)
+- [Run CodeQL daily](#run-codeql-daily)
+- [Additional checks](#additional-checks)
+  - [Automated check for markdown links](#automated-check-for-markdown-links)
+  - [Automated check for misspellings](#automated-check-for-misspellings)
+  - [Markdown lint](#markdown-lint)
+  - [Running checks against changed files only](#running-checks-against-changed-files-only)
+- [Configure component owners in contrib repositories](#configure-component-owners-in-contrib-repositories)
+- [Release automation](#release-automation)
+  - [Workflows that generate PRs](#workflows-that-generate-prs)
+  - [Prepare release branch](#prepare-release-branch)
+  - [Prepare patch](#prepare-patch)
+  - [Generate release notes from change log](#generate-release-notes-from-change-log)
+  - [Backporting PRs to a release branch](#backporting-prs-to-a-release-branch)
+- [Naming conventions](#naming-conventions)
+- [YAML style guide](#yaml-style-guide)
+
+<!-- tocstop -->
+
+</details>
+
+## Have a single required status check for PRs
+
+This avoids needing to modify branch protection required status checks as individual jobs come and go.
+
+```
+  required-status-check:
+    needs: [ aaa, bbb, ccc, ... ]
+    runs-on: ubuntu-latest
+    if: always()
+    steps:
+      - if: |
+          needs.aaa.result != 'success' ||
+          needs.bbb.result != 'success' ||
+          needs.ccc.result != 'success' ||
+          ...
+        run: exit 1
+```
+
+If you have multiple workflows that run on pull requests, there are a couple of options:
+
+* If they have the same `on` triggers, they can be merged into a single workflow.
+* Otherwise turn all but one of them into
+  [reusable workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows),
+  remove the `pull_request` trigger from the reusable workflows,
+  and call the reusable workflows from the single PR workflow.
+
+## Configure "cancel-in-progress" on PR workflows
```

Comment:
Nit: should we enable it for the CI build too?

Reply:
I sort of like to have a record of CI for each commit on `main`